File: src/realms/collages/controllers/workers.ts
import { workers } from '@services/workers/workersService.js'

export function listWorkersController() {
  return {
    workers: Array.from(workers.values()).map((worker) => ({
      name: worker.name,
      engine: worker.engine,
      version: worker.version,
      themes: worker.themes,
      available: worker.available
    }))
  }
}

File: src/main/kotlin/io/musicorum/api/realms/collages/routes/workers/Workers.kt
package io.musicorum.api.realms.collages.routes.workers

import io.ktor.server.application.*
import io.ktor.server.auth.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import io.musicorum.api.realms.collages.schemas.Worker
import io.musicorum.api.realms.collages.services.WorkersService
import io.musicorum.api.security.AuthenticationMethod
import kotlinx.serialization.Serializable
import org.koin.ktor.ext.inject

fun Route.createWorkersRoute() {
    val workersService = inject<WorkersService>()

    authenticate(AuthenticationMethod.Super, AuthenticationMethod.Client) {
        get {
            call.respond(
                WorkersResponse(
                    workersService.value.workers.map { it.toSerializable() }
                )
            )
        }
    }
}

@Serializable
data class WorkersResponse(
        val workers: List<Worker.SerializableWorker>
)